### PR TITLE
[Feature] Add hamming_distance string function for Trino/Presto [fix repo] 

### PR DESCRIPTION
**Pull request by zjoyjin**
### What problem does this PR solve?

Issue Number: subclass issue for https://github.com/apache/doris/issues/17176

Problem Summary: 
Implemented hamming_distance(string, string) in vectorized BE string functions. 
Registered the function in BE string function factory.
Added Nereids scalar function metadata (HammingDistance) and FE builtin registration.
Added FE constant-folding support for literal arguments.

Behavioral impact

Adds a new scalar string function:

-  hamming_distance(left, right)
- Returns the number of differing character positions
- Raises error when input strings have unequal length

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [X ] Unit Test
    - [X ] Manual test (add detailed scripts or steps below)
    - select hamming_distance('karolin', 'kathrin');
    - select hamming_distance('1011101', '1001001');
    - select hamming_distance('abc', 'ab'); (expect error)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [X] Yes. Adds new SQL function 'hamming_distance'.

- Does this need documentation?
    - [ ] No.
    - [X] Yes. (SQL function docs can be added in a follow-up PR if needed, but not sure if needed now)

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

File: regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy [hunk 1]
```diff
@@ -413,6 +413,17 @@ suite("test_string_function", "arrow_flight_sql") {
...
+    qt_hamming_distance2 """ select hamming_distance('1011101', '1001001'); """
+    qt_hamming_distance3 """ select hamming_distance('你好a', '你坏b'); """
+    qt_hamming_distance4 """ select hamming_distance(null, 'abc'); """
+    qt_hamming_distance5 """ select hamming_distance('abc', null); """
+
+    test {
+        sql """ select hamming_distance('abc', 'ab'); """
+        exception "hamming_distance requires strings of equal length"
+    }
+
```
Comment (Yukang-Lian) [inline regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy:429]: Missing .out file for regression tests

File: be/src/vec/functions/function_string.cpp [hunk 2]
```diff
@@ -1446,6 +1446,9 @@ void register_function_string(SimpleFunctionFactory& factory) {
     factory.register_function<FunctionMakeSet>();
     factory.register_function<FunctionExportSet>();
     factory.register_function<FunctionUnicodeNormalize>();
+    factory.register_function<FunctionOverlay>();
+    factory.register_function<FunctionStrcmp>();
```
Comment (Yukang-Lian) [inline be/src/vec/functions/function_string.cpp]: rebase onto latest master and remove these two lines from the PR.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/executable/StringArithmetic.java [hunk 3]
```diff
@@ -842,6 +842,37 @@ public static Expression strcmp(StringLikeLiteral first, StringLikeLiteral secon
         }
     }
 
+        /**
```
Comment (Yukang-Lian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/executable/StringArithmetic.java]: incorrect indentation

**Review by Yukang-Lian — CHANGES_REQUESTED** [review]
Can you add edge case tests:
    - Empty strings: select hamming_distance('', '') — should return 0
    - Same strings: select hamming_distance('abc', 'abc') — should return 0
    - Both null: select hamming_distance(null, null)

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/executable/StringArithmetic.java (hunk 3 above)
Comment (zjoyjin) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/executable/StringArithmetic.java]: hi Yukang, I've address all the issues and have update my pr

File: be/src/vec/functions/function_string.h [hunk 4]
```diff
@@ -178,6 +178,77 @@ class FunctionStrcmp : public IFunction {
...
+    size_t get_number_of_arguments() const override { return 2; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        uint32_t result, size_t input_rows_count) const override {
+        auto left_column =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:198]: dont `convert_to_full_column_if_const`, use `const_vector`, `vector_const`...

File: be/src/vec/functions/function_string.h [hunk 5]
```diff
@@ -178,6 +178,77 @@ class FunctionStrcmp : public IFunction {
...
+            if (l_char_len != r_char_len) {
+                return Status::InvalidArgument("hamming_distance requires strings of equal length");
+            }
+
+            size_t l_pos = 0;
+            size_t r_pos = 0;
+            Int64 distance = 0;
+
+            while (l_pos < l.size && r_pos < r.size) {
+                uint8_t l_step = simd::get_utf8_byte_length(static_cast<uint8_t>(l.data[l_pos]));
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:233]: for all non-utf8 inputs, we can put a fast path

File: be/src/vec/functions/function_string.h [hunk 6]
```diff
@@ -178,6 +178,77 @@ class FunctionStrcmp : public IFunction {
...
+                uint8_t r_step = simd::get_utf8_byte_length(static_cast<uint8_t>(r.data[r_pos]));
+
+                if (l_step != r_step || std::memcmp(l.data + l_pos, r.data + r_pos, l_step) != 0) {
+                    ++distance;
+                }
+
+                l_pos += l_step;
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:240]: dont need two cursor, use for-loop may easier to auto-vectorization

**Comment by zclllyybb** [comment]
continue by https://github.com/apache/doris/pull/60412
